xds/src/main/java/io/grpc/xds/RingHashLoadBalancer.java

```diff
@@ -0,0 +1,389 @@
+    subchannel.shutdown();
+    getSubchannelStateInfoRef(subchannel).set(ConnectivityStateInfo.forNonError(SHUTDOWN));
+  }
+
+  /**
+   * Converts list of {@link EquivalentAddressGroup} to {@link EquivalentAddressGroup} set and
+   * remove all attributes. The values are the original EAGs.
+   */
+  private static Map<EquivalentAddressGroup, EquivalentAddressGroup> stripAttrs(
+      List<EquivalentAddressGroup> groupList) {
+    Map<EquivalentAddressGroup, EquivalentAddressGroup> addrs =
+        new HashMap<>(groupList.size() * 2);
+    for (EquivalentAddressGroup group : groupList) {
+      addrs.put(stripAttrs(group), group);
+    }
+    return addrs;
+  }
+
+  private static EquivalentAddressGroup stripAttrs(EquivalentAddressGroup eag) {
+    return new EquivalentAddressGroup(eag.getAddresses());
+  }
+
+  private static AtomicReference<ConnectivityStateInfo> getSubchannelStateInfoRef(
+      Subchannel subchannel) {
+    return checkNotNull(subchannel.getAttributes().get(STATE_INFO), "STATE_INFO");
+  }
+
+  private final class RingHashPicker extends SubchannelPicker {
+    private final List<RingEntry> ring;
+    // shallow copy of subchannels
+    private final Map<EquivalentAddressGroup, Subchannel> pickableSubchannels;
+
+    private RingHashPicker(
+        List<RingEntry> ring, Map<EquivalentAddressGroup, Subchannel> pickableSubchannels) {
+      this.ring = ring;
+      this.pickableSubchannels = pickableSubchannels;
+    }
+
+    @Override
+    public PickResult pickSubchannel(PickSubchannelArgs args) {
+      Long requestHash = args.getCallOptions().getOption(XdsNameResolver.RPC_HASH_KEY);
+      if (requestHash == null) {
+        return PickResult.withError(Status.INTERNAL.withDescription("RPC hash not found"));
+      }
+      if (ring.isEmpty()) {
+        return PickResult.withNoResult();
+      }
+      // Find the nearest corresponding host clockwise around the ring.
+      int low = 0;
+      int high = ring.size();
+      int mid;
+      while (true) {
+        mid = (low + high) / 2;
+        if (mid == ring.size()) {
+          mid = 0;
+          break;
+        }
+        long midVal = ring.get(mid).hash;
+        long midValL = mid == 0 ? 0 : ring.get(mid - 1).hash;
+        if (requestHash <= midVal && requestHash > midValL) {
+          break;
+        }
+        if (midVal < requestHash) {
+          low = mid + 1;
+        } else {
+          high =  mid - 1;
+        }
+        if (low > high) {
+          mid = 0;
+          break;
+        }
+      }
+
+      // Try finding the first non-TRANSIENT_FAILURE connection.
+      int attempt = 0;
+      Subchannel subchannel;
+      ConnectivityStateInfo stateInfo;
+      do {
+        int index = (mid + attempt) % ring.size();
+        EquivalentAddressGroup addrKey = ring.get(index).addrKey;
+        subchannel = pickableSubchannels.get(addrKey);
+        stateInfo = getSubchannelStateInfoRef(subchannel).get();
```

Comment:
FIXME: this synchronized connectivity state read can be avoided, by _freezing_ the Picker's view of connectivity states for all subchannels when the Picker is created.

Reply:
Obsolete.